kratos/future/solving_strategies/builders/block_builder.h

```diff
@@ -0,0 +1,375 @@
+
+    /// Data type definition from sparse matrix
+    using DataType = typename TSparseMatrixType::DataType;
+
+    /// Index type definition from sparse matrix
+    using IndexType = typename TSparseMatrixType::IndexType;
+
+    /// DOF type definition
+    using DofType = typename BaseType::DofType;
+
+    /// DOF array type definition
+    using DofsArrayType = typename BaseType::DofsArrayType;
+
+    /// DOF pointer vector type definition
+    using DofPointerVectorType = typename BaseType::DofPointerVectorType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    BlockBuilder() = delete;
+
+    /// Constructor with model part
+    BlockBuilder(
+        const ModelPart &rModelPart,
+        Parameters Settings = Parameters(R"({})"))
+        : BaseType(rModelPart, Settings)
+    {
+        Parameters default_parameters( R"({
+            "name" : "block_builder",
+            "scaling_type" : "max_diagonal",
+            "echo_level" : 0
+        })");
+        Settings.ValidateAndAssignDefaults(default_parameters);
+
+        // Set scaling type
+        const std::string scaling_type = Settings["scaling_type"].GetString();
+            if (scaling_type == "no_scaling") {
+                mScalingType = ScalingType::NoScaling;
+            } else if (scaling_type == "norm_diagonal") {
+                mScalingType = ScalingType::NormDiagonal;
+            } else if (scaling_type == "max_diagonal") {
+                mScalingType = ScalingType::MaxDiagonal;
+            } else if (scaling_type == "prescribed_diagonal") {
+                mScalingType = ScalingType::PrescribedDiagonal;
+            } else {
+                KRATOS_ERROR << "Provided \'" << scaling_type << "\' is not supported." << std::endl;
+            }
+    }
+
+    virtual ~BlockBuilder() = default;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void AllocateLinearSystemArrays(
+        const typename DofsArrayType::Pointer pDofSet,
+        const typename DofsArrayType::Pointer pEffectiveDofSet,
+        LinearSystemContainer<TSparseMatrixType, TSystemVectorType> &rLinearSystemContainer) override
+    {
+        // Set up the system sparse matrix graph (note that the sparse graph will be destroyed when leaving this scope)
+        BuiltinTimer sparse_matrix_graph_time;
+        TSparseGraphType sparse_matrix_graph(pDofSet->size());
+        this->SetUpSparseMatrixGraph(sparse_matrix_graph);
+        KRATOS_INFO_IF("BlockBuilder", this->GetEchoLevel() > 0) << "Set up sparse matrix graph time: " << sparse_matrix_graph_time << std::endl;
+
+        // Set the system arrays
+        // Note that the graph-based constructor does both resizing and initialization
+        auto p_dx = Kratos::make_shared<TSystemVectorType>(sparse_matrix_graph);
+        rLinearSystemContainer.pDx.swap(p_dx);
+
+        auto p_rhs = Kratos::make_shared<TSystemVectorType>(sparse_matrix_graph);
```

Comment:
I am asking. I never understood the logic of saving the system member as shared pointers, and if we did this from scratch and still keep in, then this means that there  is a good reason.